### PR TITLE
Docusaurus documentation site

### PR DESCRIPTION
**Pull request by mongodben**
add Docusaurus documentation site to project. 

staging version of site: https://mongodben.github.io/Bluehawk

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+From the `/docs` folder:
+
+```shell
+npm start
+```
+
+This command starts a local development server and opens up a browser window. Most changes are reflected live without having to restart the server.
+
+### IMPORTANT NOE FOR FIRST BUILD
```
Comment (nlarew) [inline docs/README.md:28]:
```suggestion
### IMPORTANT NOTE FOR FIRST BUILD
```

File: docs/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+This issue occurs because the docusaurus-plugin-typedoc parser has trouble with
+a code annotation included in an imported package.
+
+You can fix this by:
+
+1. going to the file in the Bluehawk repo: `/node_modules/chevrotain/lib/chevrotain.d.ts`.
+   (_Note_: This is the `node_modules` folder in the root of the Bluehawk repo, not in the `/docs` directory.)
+2. On line 158, change `<tokType>` to `"tokType"`.
+3. Rerun `npm start` and the docs should render as expected.
```
Comment (nlarew) [inline docs/README.md:53]: Is this a known issue for the plugin or chevrotain? Is there an existing issue we can track?

File: tsconfig.json [hunk 3]
```diff
@@ -10,13 +10,10 @@
     "strict": true,
     "declaration": true,
     "declarationMap": true,
-    "resolveJsonModule": true
+    "resolveJsonModule": true,
+    "useUnknownInCatchVariables": false
```
Comment (cbush) [inline tsconfig.json:14]: Is this related?

File: docs/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+```
+
+This command generates static content into the `build` directory and can be served using any static contents hosting service.
+
+## Deployment
+
+From the `/docs` folder:
+
+```shell
+GIT_USER=<Your GitHub username> USE_SSH=true npm run deploy
```
Comment (nlarew) [inline docs/README.md:70]: Eventually it'd be great to get these as node scripts like...

in `/docs`:
```
npm run deploy
```
from the project root:
```
npm run docs:deploy
```

File: docs/README.md (hunk 2 above)
Comment (cbush) [reply docs/README.md:53]: Do we need to update chevrotain? Is there a way to exclude chevrotain from being parsed by docusaurus?

**Review by nlarew — COMMENTED** [review]
The `/docs/docs/*` directory structure feels a bit weird to me. Is this a docusaurus convention? Could we rename to something like `/docs/content/*` or something else that makes sense but reduces repetition?

**Comment by mongodben** [comment]
@nlarew having the subdirectory named `docs` is docusaurus convention. and i'd advocate we keep it as such. but we could certainly name the parent directory something else. so it's like `OTHER_NAME/docs/*`. open to ideas 😀

File: docs/README.md (hunk 4 above)
Comment (mongodben) [reply docs/README.md:70]: tru, i believe that once we get this into a github action this will be unnecessary, which'll be a lot more straightforward when the chevrotain parsing issue is fixed.

File: docs/README.md (hunk 2 above)
Comment (mongodben) [reply docs/README.md:53]: created issues for the typedoc plugin to fix the problem generally and a PR that fixes our specific issue with chevrotain. let's see what happens
* https://github.com/Chevrotain/chevrotain/pull/1726
* https://github.com/tgreyuk/typedoc-plugin-markdown/issues/276

Comment (mongodben) [reply docs/README.md:53]: got the change merged into chevrotain. though the person on their side who merged the change said it could take some time til it's in a release. 

in the meantime they recommended using [patch-package](https://github.com/ds300/patch-package) as a temporary fix. i'll investigate this.

https://github.com/Chevrotain/chevrotain/pull/1726#issuecomment-999185322

File: tsconfig.json (hunk 3 above)
Comment (mongodben) [reply tsconfig.json]: yea it's required for the typedoc docusaurus plugin to work.

if this flag isn't enabled, the plugin throws the following error when it runs:

<details>
<summary>error msg</summary>

```
Error: ../src/bluehawk/actions/copy.ts:29:62 - error TS2571: Object is of type 'unknown'.

29     const message = `Could not load stats for ${rootPath}: ${error.message}`;
                                                                ~~~~~

Error: ../src/bluehawk/actions/copy.ts:50:76 - error TS2571: Object is of type 'unknown'.

50       const message = `Failed to copy file ${filePath} to ${targetPath}: ${error.message}`;
                                                                              ~~~~~

Error: ../src/bluehawk/actions/copy.ts:111:100 - error TS2571: Object is of type 'unknown'.

111       const message = `Failed to write file ${targetPath} (based on ${parseResult.source.path}): ${error.message}`;
                                                                                                       ~~~~~

Error: ../src/bluehawk/actions/snip.ts:162:95 - error TS2571: Object is of type 'unknown'.

162       const message = `Failed to write ${targetPath} (based on ${parseResult.source.path}): ${error.message}`;
                                                                                                  ~~~~~

Error: ../src/bluehawk/bluehawk.ts:132:3 - error TS2571: Object is of type 'unknown'.

132 ${e.stack}
      ~

Error: ../src/bluehawk/bluehawk.ts:162:65 - error TS2571: Object is of type 'unknown'.

162         `falling back to plaintext parser for ${source.path}: ${error.message}`
                                                                    ~~~~~

Error: ../src/bluehawk/parser/visitor/makeCstVisitor.ts:244:37 - error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Error'.

244             jsonErrorToVisitorError(error, json, {
                                        ~~~~~

Error: ../src/cli/cli.ts:45:7 - error TS2571: Object is of type 'unknown'.

45       error.message = `Plugin '${path}' register() failed with error: ${error.message}`;
         ~~~~~

Error: ../src/cli/cli.ts:45:73 - error TS2571: Object is of type 'unknown'.

45       error.message = `Plugin '${path}' register() failed with error: ${error.message}`;
```

</details>

File: docs/README.md (hunk 2 above)
Comment (mongodben) [reply docs/README.md:53]: patch-package worked perfectly 😀

File: docs/README.md (hunk 4 above)
Comment (mongodben) [reply docs/README.md:70]: added GH actions for deployment, so this is now no longer relevant 🎉

File: tsconfig.json (hunk 3 above)
Comment (nlarew) [reply tsconfig.json]: Looks like these are errors in our (Bluehawk) code. Do you think we could update bh to cast the `unkown` errors instead (in the future perhaps)?

`unknown` is technically more correct and in general I think it's best that we stick as closely as possible to default/strict typescript.

File: docs/README.md (hunk 2 above)
Comment (cbush) [reply docs/README.md:53]: Should this note be updated then?

File: README.md [hunk 5]
```diff
@@ -1,725 +1,13 @@
...
-
-## Usage as a Module
-
-```sh
-npm install bluehawk
-```
-
-## Contributing
-
-For more information about how to run, build, or test Bluehawk yourself, see [CONTRIBUTING.md](CONTRIBUTING.md).
```
Comment (cbush) [inline README.md:725]: Why is README completely deleted?

File: docs/docs/api/classes/Bluehawk.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
+---
+id: "Bluehawk"
```
Comment (cbush) [inline docs/docs/api/classes/Bluehawk.md:2]: Do these files actually need to be committed or can they be added to gitignore and just generated on the fly (during github action)?

File: README.md (hunk 5 above)
Comment (nlarew) [reply README.md:725]: Even if we remove the actual docs/reference, I think we could keep the Example, Videos, and Install sections.

File: docs/docs/api/classes/Bluehawk.md (hunk 6 above)
Comment (mongodben) [reply docs/docs/api/classes/Bluehawk.md:2]: i think .gitignore-ing them would make sense. will make that change.

File: README.md (hunk 5 above)
Comment (mongodben) [reply README.md:725]: absolutely no idea. good catch. will add back.

File: tsconfig.json (hunk 3 above)
Comment (mongodben) [reply tsconfig.json]: i don't really understand the question, could you expand a bit?

File: docs/README.md (hunk 2 above)
Comment (mongodben) [reply docs/README.md:53]: which note? and updated to reflect what?

Comment (cbush) [reply docs/README.md:53]: Well we patched chevrotain so this isn't this part of the README now obsolete?

Comment (mongodben) [reply docs/README.md:53]: yes it should!

File: package.json [hunk 7]
```diff
@@ -61,7 +61,7 @@
     "release-it": "^14.4.1",
     "ts-jest": "^26.5.2",
     "typedoc": "^0.20.28",
-    "typescript": "^4.1.5"
+    "typescript": "^4.4.0"
```
Comment (mongodben) [inline package.json:64]: updated typescript to be able to use [`useUnknownInCatchVariables` ](https://www.typescriptlang.org/tsconfig#useUnknownInCatchVariables). not sure how it was passing tests before, as this flag was introduced in 4.4.

File: tsconfig.json (hunk 3 above)
Comment (nlarew) [reply tsconfig.json:14]: I'm mainly interested if the errors you see with `{ useUnknownInCatchVariables: true }` are something we could fix by updating the bluehawk source code. If so, I think we should fix them (i.e. [what they do on this page](https://www.typescriptlang.org/tsconfig#useUnknownInCatchVariables))

Comment (nlarew) [reply tsconfig.json:14]: I fixed this on our end in https://github.com/mongodb-university/Bluehawk/pull/76

Once we merge that can we try docusaurus with `useUnknownInCatchVariables` set to `true`?

File: README.md (hunk 5 above)
Comment (mongodben) [reply README.md:725]: update on this. if you expand the readme (i think [this link](https://github.com/mongodb-university/Bluehawk/pull/74/files/823f03266681f3691f8618901e0919b396e1afd4) will work), you should see what's left of it. it was just dramatically reduced in size b/c i moved most of the content from this 1 file to various files in the docusaurus site.

File: docs/docs/api/classes/Bluehawk.md (hunk 6 above)
Comment (mongodben) [reply docs/docs/api/classes/Bluehawk.md]: done

**Review by cbush — APPROVED** [review]

File: .github/workflows/publish-pages.yml [hunk 8]
```diff
@@ -1,27 +0,0 @@
-name: Generate Docs and Publish to GitHub Pages
```
[the remark this answers is not in the document]
Comment (mongodben) [reply .github/workflows/publish-pages.yml:1]: makes sense to me. renaming
